fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/RollCallRepository.java

```diff
@@ -0,0 +1,182 @@
+  }
+
+  /**
+   * This either updates the roll call in the repository or adds it if absent
+   *
+   * @param rollCall the roll call to update/add
+   */
+  public void updateRollCall(String laoId, RollCall rollCall, String previousId) {
+    if (laoId == null) {
+      throw new IllegalArgumentException("Lao id is null");
+    }
+    if (rollCall == null) {
+      throw new IllegalArgumentException("Roll call is null");
+    }
+    Log.d(TAG, "Adding new roll call on lao " + laoId + " : " + rollCall);
+
+    // Retrieve Lao data and add the roll call to it
+    getLaoRollCalls(laoId).update(new RollCall(rollCall), previousId);
+  }
+
+  /**
+   * This provides an observable of a roll call that triggers an update when modified
+   *
+   * @param laoId the id of the Lao
+   * @param persistentId the persistent id of the roll call
+   * @return the observable wrapping the wanted roll call
+   * @throws UnknownRollCallException if no roll call with the provided id could be found
+   */
+  public Observable<RollCall> getRollCallObservable(String laoId, String persistentId)
+      throws UnknownRollCallException {
+    return getLaoRollCalls(laoId).getRollCallObservable(persistentId);
+  }
+
+  public RollCall getRollCallWithPersistentId(String laoId, String persistentId)
+      throws UnknownRollCallException {
+    return getLaoRollCalls(laoId).getRollCallWithPersistentId(persistentId);
+  }
+
+  public RollCall getRollCallWithId(String laoId, String rollCallId)
+      throws UnknownRollCallException {
+    return getLaoRollCalls(laoId).getRollCallWithId(rollCallId);
+  }
+
+  /**
+   * @param laoId the id of the Lao whose roll calls we want to observe
+   * @return an observable set of ids who correspond to the set of roll calls published on the given
+   *     lao
+   */
+  public Observable<Set<String>> getRollCallsObservableInLao(String laoId) {
+    return getLaoRollCalls(laoId).getRollCallsSubject();
+  }
+
+  /**
+   * Returns the set of all attendees who have ever attended a roll call in the lao
+   *
+   * @param laoId the id of the considered lao
+   * @return the set of all attendees who have ever attended a roll call in the lao
+   */
+  public Set<PublicKey> getAllAttendeesInLao(String laoId) {
+    return getLaoRollCalls(laoId).getAllAttendees();
+  }
+
+  public RollCall getLastClosedRollCall(String laoId) throws NoRollCallException {
+    return getLaoRollCalls(laoId).rollCallByPersistentId.values().stream()
+        .filter(RollCall::isClosed)
+        .max(Comparator.comparing(RollCall::getEnd))
+        .orElseThrow(() -> new NoRollCallException(laoId));
+  }
+
+  @NonNull
+  private synchronized LaoRollCalls getLaoRollCalls(String laoId) {
+    return rollCallsByLao.computeIfAbsent(laoId, lao -> new LaoRollCalls());
+  }
+
+  private static final class LaoRollCalls {
+    private final Map<String, RollCall> rollCallByPersistentId = new HashMap<>();
+    private final Map<String, RollCall> rollCallById = new HashMap<>();
```

Comment:
Another idea : instead of having 2 maps of size N, we could have 1 map of size 2n with each roll calls (in fact only roll calls whose id and persistentId differ, that is the opened and closed ones) are stored twice. I don't know if the hash collision could be a problem though 🤔 
Perhahps the alias idea is better, I'll implement it.

Reply:
I think your idea starts to get complex once you try to update a rollcall at a certain id and you need to keep track of all other ids